#**Information Retrieval CSF469**
##Lab Session - 2
## Date - 31/01/2024
##<font color ='GREEN'>Marks: 10</font>
In this lab session, we implement label sensitive boolean query retrivel.

## Students have to write their code in blank spaces marked under <font color = "RED">TO-DO</font> section


You are given documents relating to two companies, A and B. You are required to upgrade the standard boolean retrieval system to improve the accuracy of retrieving documents given the labels that if they belong to the company A or B. For eg. if the query consists of the name ‘A’, the query should only return documents belonging to company A even if some documents of company B also match the full query.

In [ ]:
import pandas as pd

In [ ]:
dataset = pd.read_csv('IRdata.csv')
# Dataset is now stored in a Pandas Dataframe

In [ ]:
print(dataset)

                                                 title     label
0    Avantel Limited Announces Resignation of Ebv R...   Avantel
1    480% Returns From 1-Year Low: Multibagger Defe...   Avantel
2    Multibagger defence stock rises 5% on Rs 68 cr...   Avantel
3    Avantel Ltd receives order worth Rs. 67.92 cro...   Avantel
4    Rs 11 to Rs 129: This defence stock turned int...   Avantel
..                                                 ...       ...
157  Titagarh Rail Systems rolls out new diving sup...  Titagarh
158  Titagarh Rail shares fall 10% from record high...  Titagarh
159  Rs 49 to Rs 813: This railway stock turned int...  Titagarh
160  Škoda Group secures €732m contract with Trenit...  Titagarh
161  Titagarh Rail Share Price: Stock at record hig...  Titagarh

[162 rows x 2 columns]


In [ ]:
df = dataset.iloc[:, 0]
df2 = dataset.iloc[:, -1]

In [ ]:
print(df)
df2 = df2.to_frame()
print(df2)

0      Avantel Limited Announces Resignation of Ebv R...
1      480% Returns From 1-Year Low: Multibagger Defe...
2      Multibagger defence stock rises 5% on Rs 68 cr...
3      Avantel Ltd receives order worth Rs. 67.92 cro...
4      Rs 11 to Rs 129: This defence stock turned int...
                             ...                        
157    Titagarh Rail Systems rolls out new diving sup...
158    Titagarh Rail shares fall 10% from record high...
159    Rs 49 to Rs 813: This railway stock turned int...
160    Škoda Group secures €732m contract with Trenit...
161    Titagarh Rail Share Price: Stock at record hig...
Name: title, Length: 162, dtype: object
        label
0     Avantel
1     Avantel
2     Avantel
3     Avantel
4     Avantel
..        ...
157  Titagarh
158  Titagarh
159  Titagarh
160  Titagarh
161  Titagarh

[162 rows x 1 columns]


In [ ]:
import nltk
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import WordNetLemmatizer
import string

In [ ]:
nltk.download('punkt')
nltk.download('stopwords')
nltk.download('wordnet')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


True

In [ ]:
stop_words = set(stopwords.words('english'))
lemmatizer = WordNetLemmatizer()

In [ ]:
def preprocess_text(text):
    # Tokenize the text
    tokens = word_tokenize(text)

    # Remove punctuation and convert to lower case
    tokens = [word.lower() for word in tokens if word.isalnum()]
    # tokens = [word.lower() for word in tokens if word.isalpha()]

    # keep numeric as well
    #TO-DO    #isalnum()
    # Remove stop words
    tokens = [word for word in tokens if word not in stop_words]#TO-DO

    # Lemmatize each word
    # lemmatized_tokens = [lemmatizer.lemmatize(word) for word in tokens]

    return tokens

    # remove lemmatizer

In [ ]:
processed_data = df.apply(preprocess_text)

In [ ]:
print(processed_data)
# after preprocesing rs129 whould be one token how??

0      [avantel, limited, announces, resignation, ebv...
1      [480, returns, low, multibagger, defence, stoc...
2      [multibagger, defence, stock, rises, 5, rs, 68...
3      [avantel, ltd, receives, order, worth, rs, cro...
4      [rs, 11, rs, 129, defence, stock, turned, mult...
                             ...                        
157    [titagarh, rail, systems, rolls, new, diving, ...
158    [titagarh, rail, shares, fall, 10, record, hig...
159    [rs, 49, rs, 813, railway, stock, turned, mult...
160    [škoda, group, secures, contract, trenitalia, ...
161    [titagarh, rail, share, price, stock, record, ...
Name: title, Length: 162, dtype: object


In [ ]:
# Define the inverted index dictionary
inverted_index = {}

#to-do
# Iterate through each document and its processed tokens
for doc_id, tokens in enumerate(processed_data):
    for token in tokens:
        # If the token is not already in the dictionary, add it
        if token not in inverted_index:
            inverted_index[token] = set()

        # Add the document ID to the set of documents for this token
        inverted_index[token].add(doc_id)

In [ ]:
for word, doc_ids in inverted_index.items():
    print(f"{word}: {list(doc_ids)}")

avantel: [0, 3, 7, 10, 14, 15, 18, 19, 20, 21, 23, 24, 26, 28, 29, 31, 35, 40, 41, 45, 46, 47, 48, 49, 51, 52, 55, 56, 57, 58, 59, 60, 61, 62]
limited: [0, 128, 98, 43, 49, 150, 151, 56]
announces: [0]
resignation: [0]
ebv: [0]
ramana: [0]
gupta: [0]
independent: [0]
director: [0]
member: [0]
audit: [0]
480: [1]
returns: [1, 130, 36, 39, 43, 78]
low: [1, 13]
multibagger: [1, 2, 130, 4, 5, 6, 138, 11, 13, 17, 154, 30, 159, 32, 36, 37, 39, 43, 44, 54, 72, 77, 127]
defence: [1, 2, 4, 5, 6, 27]
stock: [1, 2, 130, 4, 5, 6, 138, 11, 12, 13, 16, 146, 20, 22, 151, 25, 154, 27, 30, 159, 32, 33, 34, 161, 36, 37, 38, 39, 44, 45, 49, 54, 57, 66, 71, 72, 73, 77, 78, 90, 93, 127]
less: [1]
rs: [1, 2, 3, 4, 129, 6, 136, 137, 138, 14, 15, 18, 147, 148, 153, 29, 159, 35, 42, 48, 52, 55, 56, 60, 72, 87, 90, 93, 94, 95, 97, 109, 117, 127]
10: [1, 39, 85, 53, 26, 158]
away: [1]
new: [1, 34, 66, 43, 44, 45, 80, 123, 157]
high: [1, 2, 131, 4, 135, 11, 12, 22, 158, 161, 44, 45, 52, 55, 72, 82, 89, 102, 104, 

In [ ]:
def boolean_query(query, inverted_index, df):
    # Tokenize the query into words and operators
    query_parts = query.lower().split()

    # Separate words and operators
    # words = [token for token in query_parts if token not in ["and", "or"]]
    words = preprocess_text(query)
    operators = [token for token in query_parts if token in ["and", "or"]]

    # Initialize result with the document set of the first word
    # Assuming words are already preprocessed
    result_set = set(inverted_index.get(words[0], []))


    #todo iterate
    # Iterate over the words and apply boolean operations
    for i, operator in enumerate(operators):
        next_word_docs = set(inverted_index.get(words[i + 1], []))

        if operator == "and":
            result_set &= next_word_docs
        elif operator == "or":
            result_set |= next_word_docs

    # Fetch the document data for each ID in the result set
    document_data = df.iloc[list(result_set)]
    return result_set
    # return result_set

In [ ]:
def label_enhanced_boolean_query(query, inverted_index, df, df2):
    # Convert query to lower case and split into components
    query_parts = query.lower().split()

    # Identify positions of 'avantel' and 'titagarh' in the query
    avantel_pos = query_parts.index('avantel') if 'avantel' in query_parts else -1
    titagarh_pos = query_parts.index('titagarh') if 'titagarh' in query_parts else -1

    # Remove 'avantel' or 'titagarh' and the adjacent word
    for label in ['avantel', 'titagarh']:
        while label in query_parts:
            label_pos = query_parts.index(label)

            # Check if the label is the last word
            if label_pos == len(query_parts) - 1:
                # If label is not the first word, remove the previous word
                if label_pos > 0:
                    query_parts.pop(label_pos - 1)
                # Remove the label
                query_parts.pop(label_pos - 1)
            else:
                # If label is not the last word, remove the next word first, then the label
                query_parts.pop(label_pos + 1)  # Remove next word
                query_parts.pop(label_pos)      # Remove the label

    print(query_parts)
    # If both or none of the labels were originally present, do a normal boolean search
    if (avantel_pos != -1 and titagarh_pos != -1):
      num_rows = df.shape[0]
      all_index = list(range(num_rows))
      return all_index
    elif not query_parts:
      num_rows = df.shape[0]
      all_index = list(range(num_rows))
      result_set = all_index
    elif avantel_pos == titagarh_pos == -1:
        return boolean_query(' '.join(query_parts), inverted_index, df)

    # Perform boolean search on the modified query
    else:
      result_set = boolean_query(' '.join(query_parts), inverted_index, df)

    # Filter the results based on the label if only one label is present
    label_filter = None
    if avantel_pos != -1:
        label_filter = 'avantel'
    elif titagarh_pos != -1:
        label_filter = 'titagarh'

    if label_filter:
        filtered_results = dataset[dataset['label'].str.lower() == label_filter]
        valid_indices = [index for index in result_set if index in filtered_results.index]
        return valid_indices

    return result_set

In [ ]:
query = input("Enter a boolean query with optional labels (e.g., 'word1 AND word2 OR avantel word3'): ")
resulting_set = label_enhanced_boolean_query(query, inverted_index, df, df2)
mask = df.index.isin(resulting_set)
filtered_df = df[mask]
print(f"Documents satisfying '{query}':\n{filtered_df}")

Enter a boolean query with optional labels (e.g., 'word1 AND word2 OR avantel word3'): crore or sales and titagarh
['crore', 'or', 'sales']
Documents satisfying 'crore or sales and titagarh':
[136, 137, 142, 143, 146, 87, 88, 90, 92, 94, 95, 97]


In [ ]:
query = input("Enter a boolean query with optional labels (e.g., 'word1 AND word2 OR avantel word3'): ")
resulting_set = boolean_query(query, inverted_index, df)
print(f"Documents satisfying '{query}':\n{resulting_set}")

Enter a boolean query with optional labels (e.g., 'word1 AND word2 OR avantel word3'): a
Documents satisfying 'a':
Invalid query format.


#Lab 2
###Answer to query 1:
Documents satisfying 'avantel and order or book':
{35, 3, 14, 47, 15, 58, 60, 158}
###Answer to query 2:
Documents satisfying 'Titagarh and intrnational or business or client':
{2, 130, 4, 9, 137, 138, 139, 140, 147, 149, 23, 152, 25, 26, 154, 30, 158, 159, 42, 45, 57, 60, 68, 69, 72, 82, 85, 89, 91, 94, 95, 109, 110, 112, 115, 117, 122, 124, 125, 127}



# Lab 3
###Answer to query 1:
Documents satisfying 'avantel or share or defence':
[1, 2, 4, 5, 6, 9, 12, 20, 21, 24, 27, 57, 31]
###Answer to query 2:
Documents satisfying 'crore or sales and titagarh':
[136, 137, 142, 143, 146, 87, 88, 90, 92, 94, 95, 97]


